src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -48,413 +46,1245 @@ internal static class ThreadPoolGlobals
+                        if (!missedSteal)
+                        {
+                            missedSteal = Volatile.Read(ref this[position - 1].SequenceNumber) != (position + _slotsMask);
+                        }
+
+                        // Read the sequence number for the slot.
+                        ref Slot slot = ref this[position];
+                        int sequenceNumber = slot.SequenceNumber;
+
+                        // Check if the slot is considered Full in the current generation.
+                        int diff = sequenceNumber - position;
+                        if (diff == Full)
+                        {
+                            // Reserve the slot for Dequeuing.
+                            if (Interlocked.CompareExchange(ref slot.SequenceNumber, position + Change, sequenceNumber) == sequenceNumber)
+                            {
+                                object? item;
+                                var enqPos = _queueEnds.Enqueue;
+
+                                if (enqPos - position < RichCount ||
+                                    // take from the rich and give to the needy!!
+                                    // NB: "this" is a sentinel for a failed robbing attempt
+                                    (item = TryRob(position, enqPos)) == this)
+                                {
+                                    _queueEnds.Dequeue = position + 1;
+                                    item = slot.Item;
+                                    slot.Item = null;
+                                }
+
+                                // unlock the slot for enqueuing by making the slot empty in the next generation
+                                Volatile.Write(ref slot.SequenceNumber, position + 1 + _slotsMask);
+
+                                if (item == null)
+                                {
+                                    // the item was removed, so we have nothing to return.
+                                    // this is not a lost race though, so must continue.
+                                    continue;
+                                }
+
+                                return item;
+                            }
+                        }
+                        else if (diff == 0)
                         {
-                            if (taken)
-                                m_foreignLock.Exit(useMemoryBarrier: false);
+                            // reached an empty slot
+                            // since full slots are contiguous, finding an empty slot means that
+                            // for our purposes and for the moment in time the segment is empty
+                            return null;
                         }
 
+                        // contention with other thread
+                        // must check this segment again later
                         missedSteal = true;
+                        return null;
```

Comment:
Why give up on contention? The path taken with `missedSteal = true` is a lot more expensive than retrying on contention. I would suggest retrying on contention (with proper spin-waiting) and return only when the info indicates that the queue is empty.

Reply:
Nevermind, I see that the retry is being done in `Dispatch` over global and all local queues again